Zhylka/legacy chips rewriting

MaxZhylka opened the pull request:
Replaced MatLegacyChipModule,  replaced all chip-lists by chip-set. In the CreateDescriptionForm component, specifically updated mat-chip-list to mat-chip-grid to resolve test failures

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.html:
```diff
@@ -60,7 +60,7 @@
       {{ 'FORMS.PLACEHOLDERS.KEYWORDS_END' | translate }}</span
     >
-    <mat-chip-list #chipList>
+    <mat-chip-grid #chipGrid>
```
It is better to use mat-chip-grid together with the input to follow the correct [interaction pattern](https://v16.material.angular.io/components/chips/overview#text-entry)

humanmantis commented on src/app/shell/personal-cabinet/parent/create-child/child-form/child-form.component.ts:
```diff
@@ -34,8 +34,9 @@ export class ChildFormComponent implements OnInit, OnDestroy {
   private select: MatSelect;
-  @ViewChild('chipList')
-  private chipList: MatChipList;
+
```
remove empty  row

MaxZhylka commented:
I am not sure if ngModel needs to be set on mat-chip-grid (as shown in the docs), because we use reactive forms. So, I removed ngModel from the input and fixed the issue where the input remained disabled after adding 5 elements and then removing one.

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.html:
```diff
@@ -53,31 +53,30 @@
+      </mat-chip-row>
       <input
-        [ngClass]="{ 'keywords-input': !keyWord && keyWords.length < 1 }"
+        [ngClass]="{  'keywords-input': !keyWordsCtrl.value && keyWords$.value.length < 1  }"
```
remove extra whitespace after `{ `and before `}`

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.ts:
```diff
@@ -40,12 +40,13 @@ export class CreateDescriptionFormComponent implements OnInit, OnDestroy {
   public disabilityOptionRadioBtn: FormControl = new FormControl(false);
   public disabledKeyWordsInput = false;
 
+  public separatorKeysCodes = [13, 188]; // Enter and Comma
```
Import keycodes from `@angular/cdk/keycodes`

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.ts:
```diff
@@ -74,18 +75,29 @@ export class CreateDescriptionFormComponent implements OnInit, OnDestroy {
     this.keyWordsListener();
+
+    this.keyWords$.pipe(takeUntil(this.destroy$)).subscribe((keyWords) => {
+      this.disabledKeyWordsInput = keyWords.length >= 5;
```
create a private read-only property for max keywords instead of a magic number

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.ts:
```diff
@@ -40,12 +40,13 @@ export class CreateDescriptionFormComponent implements OnInit, OnDestroy {
   public keyWordsCtrl: FormControl = new FormControl('', Validators.required);
 
-  public keyWords: string[] = [];
+  public keyWords$: BehaviorSubject<string[]> = new BehaviorSubject<string[]>([]);
```
Why do we need a subject in this case?
We can simply do checks inside the method or create a separate method to switch the input state

humanmantis commented on src/app/shell/personal-cabinet/provider/create-workshop/create-description-form/create-description-form.component.ts:
```diff
@@ -74,18 +75,29 @@ export class CreateDescriptionFormComponent implements OnInit, OnDestroy {
-      if (this.keyWords.length) {
-        this.DescriptionFormGroup.get('keyWords').setValue([...this.keyWords]);
+    const keyWords = this.keyWords$.value;
+    const index = keyWords.indexOf(word);
```
`keyWords.includes()`
